### PR TITLE
added final cobre scripts

### PR DESCRIPTION
**Pull request by clarkenj**
These are a bit hacky, using a fake site variable, but when I tried to re-write the scripts without it I kept getting the `OperationalError: database is locked`. Will work on it for other datasets. 

Also I increased the times since when I did a test run with MIST and the AFC flag it took nearly 4 hours.

### CODE REVIEW

File: cobre/connectome_slurm.bash [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+#SBATCH --account=rrg-pbellec
+#SBATCH --output=/lustre04/scratch/nclarke/logs/%x_%A.out
+#SBATCH --error=/lustre04/scratch/nclarke/logs/%x_%A.err
+#SBATCH --cpus-per-task=1
+
+
+GIGA_CONNECTOME_VERSION=0.4.1
+SITE=cobre_test
```
Comment (htwangtw) [inline cobre/connectome_slurm.bash:9]:
```suggestion
```

File: cobre/connectome_slurm.bash [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+GIGA_CONNECTOME=/home/${USER}/projects/rrg-pbellec/containers/giga_connectome-${GIGA_CONNECTOME_VERSION}.simg
+FMRIPREP_DIR=/lustre04/scratch/${USER}/${DATASET}/COBRE/fmriprep-20.2.7lts/derivatives
+CONNECTOME_OUTPUT=/lustre04/scratch/${USER}/${DATASET}_connectomes-${GIGA_CONNECTOME_VERSION}/
+
+WORKINGDIR=${CONNECTOME_OUTPUT}/working_directory/${SITE}
```
Comment (htwangtw) [inline cobre/connectome_slurm.bash:14]:
```suggestion
WORKINGDIR=${CONNECTOME_OUTPUT}/working_directory/
```

File: cobre/connectome_slurm.bash [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+module load apptainer/1.1.8
+
+mkdir -p $WORKINGDIR
+
+echo "${FMRIPREP_DIR}"
+if [ -d "${FMRIPREP_DIR}" ]; then
+    mkdir -p ${WORKINGDIR}
+	mkdir -p ${SLURM_TMPDIR}
+	mkdir -p ${CONNECTOME_OUTPUT}/${SITE}
```
Comment (htwangtw) [inline cobre/connectome_slurm.bash:24]:
```suggestion
	mkdir -p ${CONNECTOME_OUTPUT}
```

File: cobre/connectome_slurm.bash [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+		--bind ${WORKINGDIR}:/data/working \
+		${GIGA_CONNECTOME} \
+		-w /data/working \
+		--atlas ${ATLAS} \
+		--denoise-strategy ${STRATEGY} \
+		/data/input \
+		/data/output \
+		group
+	exitcode=$?  # catch exit code
+	if [ $exitcode -eq 0 ] ; then rsync -rltv --info=progress2 ${SLURM_TMPDIR}/*.h5 ${CONNECTOME_OUTPUT}/${SITE} ; fi
```
Comment (htwangtw) [inline cobre/connectome_slurm.bash:39]:
```suggestion
	if [ $exitcode -eq 0 ] ; then rsync -rltv --info=progress2 ${SLURM_TMPDIR}/*.h5 ${CONNECTOME_OUTPUT} ; fi
```

File: cobre/connectome_slurm.bash [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+else
+    echo "no preprocessed data for ${SITE}"
```
Comment (htwangtw) [inline cobre/connectome_slurm.bash:41]:
```suggestion
    echo "no preprocessed data for ${DATASET}"
```

File: cobre/submit_connectomes.sh [hunk 6]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+#SBATCH --account=rrg-pbellec
+GIGA_CONNECTOME_VERSION=0.4.1
+DATASET="cobre"
+
+FMRIPREP_PATH="/lustre04/scratch/${USER}/${DATASET}/COBRE/fmriprep-20.2.7lts/derivatives"
+COBRE_CONNECTOME=/lustre04/scratch/${USER}/${DATASET}_connectomes-${GIGA_CONNECTOME_VERSION}
+
+site=cobre_test
```
Comment (htwangtw) [inline cobre/submit_connectomes.sh:9]:
```suggestion
```

File: cobre/submit_connectomes.sh [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    for atlas in $ATLASES; do
+            mem=8G
+            time="6:00:00"
+
+            if [ "${atlas}" == "DiFuMo" ]; then
+                time="10:00:00"
+                mem=32G
+            fi
+
+            if [ ! -f "${COBRE_CONNECTOME}/${site}/atlas-${atlas}_desc-${strategy}.h5" ]; then
```
Comment (htwangtw) [inline cobre/submit_connectomes.sh:23]:
```suggestion
            if [ ! -f "${COBRE_CONNECTOME}/atlas-${atlas}_desc-${strategy}.h5" ]; then
```

File: cobre/submit_connectomes.sh [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+                echo "Submitting ${site} ${atlas} ${strategy}"
```
Comment (htwangtw) [inline cobre/submit_connectomes.sh:24]:
```suggestion
                echo "Submitting ${atlas} ${strategy}"
```

File: cobre/submit_connectomes.sh [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+                sbatch \
+                    --time=${time} --mem-per-cpu=${mem} \
+                    --job-name=${DATASET}_${site}_${atlas}_${strategy} \
```
Comment (htwangtw) [inline cobre/submit_connectomes.sh:27]:
```suggestion
                    --job-name=${DATASET}_${atlas}_${strategy} \
```

File: cobre/submit_connectomes.sh [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+                    --export=DATASET="${DATASET}",ATLAS="${atlas}",STRATEGY="${strategy}" \
+                    ./connectome_slurm.bash
+            else
+                echo "Skipping ${site} ${atlas} ${strategy}"
```
Comment (htwangtw) [inline cobre/submit_connectomes.sh:31]:
```suggestion
                echo "Skipping ${atlas} ${strategy}"
```

**Review by htwangtw — APPROVED** [review]
If it works it works at this point. Just some comments to show how to remove all the site related thing that are basically just place holders.
If you ever want to test anything related to where a file is saved etc, replace the actual operation with a simple text file but keep the rest. That should be a quick and low stake way to test the thing

File: cobre/connectome_slurm.bash [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+	echo "=========${STRATEGY}========="
+	echo "${ATLAS}"
+	apptainer run \
+		--bind ${FMRIPREP_DIR}:/data/input \
+		--bind ${SLURM_TMPDIR}:/data/output \
+		--bind ${WORKINGDIR}:/data/working \
+		${GIGA_CONNECTOME} \
+		-w /data/working \
+		--atlas ${ATLAS} \
+		--denoise-strategy ${STRATEGY} \
```
[the remark this answers is not in the document]
Comment (clarkenj) [reply cobre/connectome_slurm.bash:34]: OOPS, I must have forgot to add it after testing with a different script. I will re-run. Thanks for spotting!

**Comment by clarkenj** [comment]
> If it works it works at this point. Just some comments to show how to remove all the site related thing that are basically just place holders. If you ever want to test anything related to where a file is saved etc, replace the actual operation with a simple text file but keep the rest. That should be a quick and low stake way to test the thing

Thank you!! I will try with your suggestions (and the AFC flag this time..) and re-run.

**Comment by clarkenj** [comment]
I tried all the [edits](https://github.com/SIMEXP/giga_preprocess2/pull/22/commits/b47c6661e74ad5bee0c17590ea5be5c24346f606) as suggested, but now I get the below error. Sorry, I know this is like deja vu. I have triple checked that it is the correct path, and it's fine when I run the connectome generation just once with [this](https://github.com/SIMEXP/giga_preprocess2/pull/22/commits/e765bc3f6405b71fc1d9604f91de4cf81133f366) script. Can you see anything I might have done wrong? Thank you.

```
Traceback (most recent call last):
  File "/usr/local/bin/giga_connectome", line 8, in <module>
    sys.exit(main())
  File "/code/giga_connectome/run.py", line 111, in main
    workflow(args)
  File "/code/giga_connectome/workflow.py", line 74, in workflow
    subj_data, fmriprep_bids_layout = utils.get_bids_images(
  File "/code/giga_connectome/utils.py", line 21, in get_bids_images
    layout = BIDSLayout(
  File "/usr/local/lib/python3.9/site-packages/bids/layout/layout.py", line 128, in __init__
    root, description = validate_root(root, validate)
  File "/usr/local/lib/python3.9/site-packages/bids/layout/validation.py", line 70, in validate_root
    raise ValueError("BIDS root does not exist: %s" % root)
ValueError: BIDS root does not exist: /inputs
```

**Comment by clarkenj** [comment]
I just had a thought that I did not add --reindex-bids. Unsure if it would affect that but will try.

**Comment by clarkenj** [comment]
With the `--reindex-bids` flag I get the `OperationalError: database is locked` error. Will keep investigating

**Comment by htwangtw** [comment]
> With the `--reindex-bids` flag I get the `OperationalError: database is locked` error. Will keep investigating

This might be the resaon of #24 this flag will make you constantly remove and reindex the BIDS layout

**Comment by htwangtw** [comment]
Do exactly the following:
- remove `layout_index.sqlite`
- Run one job on the dataset
- Submit the rest of the jobs in parallel

**Comment by clarkenj** [comment]
@htwangtw previous issues are resolved, thanks for your help. Final scripts added and re-organised.

**Review by htwangtw — APPROVED** [review]
This is awesome and comprehensive

**Pull request merged**